File: typescript/src/problems101_200/rotate-array.ts
class Problem {
  rotate(nums: number[], k: number) {
    throw new Error(`Not implemented (${nums}, ${k}).`);
  }
}

class Default extends Problem {
  rotate(nums: number[], k: number) {
    if (k === 0 || k % nums.length === 0) {
      return;
    }
    if (k > nums.length) {
      k %= nums.length;
    }
    this.reverse(nums, 0, nums.length - 1);
    this.reverse(nums, 0, k - 1);
    this.reverse(nums, k, nums.length - 1);
  }

  reverse(nums: number[], left: number, right: number) {
    while (left < right) {
      const temp = nums[left];
      nums[left] = nums[right];
      nums[right] = temp;
      left++;
      right--;
    }
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;

File: python/src/problems101_200/rotate_array.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def rotate(self, nums: list[int], k: int) -> None:
        pass


class Default(Problem):
    @override
    def rotate(self, nums: list[int], k: int) -> None:
        if k == 0 or k % len(nums) == 0:
            return
        if k > len(nums):
            k %= len(nums)
        self.reverse(nums, 0, len(nums) - 1)
        self.reverse(nums, 0, k - 1)
        self.reverse(nums, k, len(nums) - 1)

    def reverse(self, nums: list[int], left: int, right: int) -> None:
        while left < right:
            temp: int = nums[left]
            nums[left] = nums[right]
            nums[right] = temp
            left += 1
            right -= 1


SOLUTIONS: list[Problem] = [Default()]
